File: packages/mlforge_sdk/mlforge_sdk/datasets.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict

from .http import HttpClient

class Dataset(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    name: str
    format: Optional[str] = None
    images: int = 0
    classes: int = 0
    size_label: str = "0 MB"


class DatasetJob(BaseModel):
    model_config = ConfigDict(extra="ignore")
    id: str
    type: str
    status: str
    dataset_id: str
    dataset_name: str
    progress: float = 0.0
    message: str = ""
    error: Optional[str] = None


class ImportResponse(BaseModel):
    model_config = ConfigDict(extra="ignore")
    job_id: str
    dataset_id: str
    status: str
    message: str

class DatasetClient:
    def __init__(self, http: HttpClient):
        self._http = http

    def list(
        self,
        *,
        task: Optional[str] = None,
        format: Optional[str] = None,
        source: Optional[str] = None,
        status: Optional[str] = None,
        search: Optional[str] = None,
        starred: Optional[bool] = None,
        limit: int = 100,
        offset: int = 0,
    ) -> List[Dataset]:
        params: Dict[str, Any] = {
            "task": task,
            "format": format,
            "source": source,
            "status": status,
            "search": search,
            "starred": starred,
            "limit": limit,
            "offset": offset,
        }
        data = self._http.get("/datasets", params=params)
        return [Dataset(**d) for d in data]

    def get(self, dataset_id: str) -> Dataset:
        data = self._http.get(f"/datasets/{dataset_id}")
        return Dataset(**data)

    def import_dataset(self, dataset_id: str, payload: Dict[str, Any]) -> ImportResponse:
        data = self._http.post(f"/datasets/{dataset_id}/import", json_body=payload)
        return ImportResponse(**data)

    def list_jobs(self, *, limit: int = 50) -> List[DatasetJob]:
        data = self._http.get("/datasets/jobs", params={"limit": limit})
        return [DatasetJob(**j) for j in data]

    def get_job(self, job_id: str) -> DatasetJob:
        data = self._http.get(f"/datasets/jobs/{job_id}")
        return DatasetJob(**data)

    def stop_job(self, job_id: str) -> Dict[str, Any]:
        return self._http.post(f"/datasets/jobs/{job_id}/stop")

    def pause_job(self, job_id: str) -> Dict[str, Any]:
        return self._http.post(f"/datasets/jobs/{job_id}/pause")

    def resume_job(self, job_id: str) -> Dict[str, Any]:
        return self._http.post(f"/datasets/jobs/{job_id}/resume")

    def toggle_star(self, dataset_id: str) -> Dict[str, Any]:
        return self._http.post(f"/datasets/{dataset_id}/star")

    def delete(self, dataset_id: str, *, delete_files: bool = False) -> Dict[str, Any]:
        return self._http.delete(f"/datasets/{dataset_id}", params={"delete_files": delete_files})

    def search_roboflow(
        self,
        api_key: str,
        query: str,
        workspace: Optional[str] = None,
        page: int = 1,
        page_size: int = 20,
    ) -> List[Dataset]:
        payload = {
            "api_key": api_key,
            "query": query,
            "workspace": workspace,
            "page": page,
            "page_size": page_size,
        }
        data = self._http.post("/datasets/search/roboflow", json_body=payload)
        return [Dataset(**d) for d in data]

    def sync_roboflow(self, api_key: str, workspace: str) -> Dict[str, Any]:
        params = {"api_key": api_key, "workspace": workspace}
        return self._http.post("/datasets/sync/roboflow", params=params)

File: packages/mlforge_js_sdk/src/datasets.ts
import type { HttpClient } from "./http.js";

export interface Dataset {
  id: string;
  name: string;
  format?: string;
  images?: number;
  classes?: number;
  size_label?: string;
}

export interface DatasetJob {
  id: string;
  type: string;
  status: string;
  dataset_id: string;
  dataset_name: string;
  progress?: number;
  message?: string;
  error?: string;
}

export interface DatasetAnalytics {
  dataset_id: string;
  healthScore?: number;
  split?: Record<string, number>;
  qualityIssues?: Record<string, number>;
  classDistribution?: Array<Record<string, unknown>>;
}

export interface ImportOptions {
  project_id: string;
  [key: string]: unknown;
}

export class DatasetsClient {
  constructor(private http: HttpClient) {}

  async list(opts?: {
    task?: string;
    format?: string;
    search?: string;
    limit?: number;
    offset?: number;
  }): Promise<Dataset[]> {
    return this.http.get<Dataset[]>("/datasets", {
      task: opts?.task,
      format: opts?.format,
      search: opts?.search,
      limit: opts?.limit ?? 100,
      offset: opts?.offset ?? 0,
    } as Record<string, string | number | boolean>);
  }

  async get(datasetId: string): Promise<Dataset> {
    return this.http.get<Dataset>(`/datasets/${datasetId}`);
  }

  async getAnalytics(datasetId: string): Promise<DatasetAnalytics> {
    return this.http.get<DatasetAnalytics>(`/datasets/${datasetId}/analytics`);
  }

  async import(datasetId: string, opts: ImportOptions): Promise<{ job_id: string; dataset_id: string }> {
    return this.http.post<{ job_id: string; dataset_id: string }>(
      `/datasets/${datasetId}/import`,
      opts
    );
  }

  async listJobs(opts?: { limit?: number }): Promise<DatasetJob[]> {
    return this.http.get<DatasetJob[]>("/datasets/jobs", { limit: opts?.limit ?? 50 });
  }

  async getJob(jobId: string): Promise<DatasetJob> {
    return this.http.get<DatasetJob>(`/datasets/jobs/${jobId}`);
  }

  async stopJob(jobId: string): Promise<void> {
    return this.http.post<void>(`/datasets/jobs/${jobId}/stop`);
  }

  async pauseJob(jobId: string): Promise<void> {
    return this.http.post<void>(`/datasets/jobs/${jobId}/pause`);
  }

  async resumeJob(jobId: string): Promise<void> {
    return this.http.post<void>(`/datasets/jobs/${jobId}/resume`);
  }

  async delete(datasetId: string, opts?: { delete_files?: boolean }): Promise<void> {
    return this.http.delete<void>(`/datasets/${datasetId}`, {
      delete_files: opts?.delete_files ?? false,
    });
  }

  async searchRoboflow(opts: {
    api_key: string;
    query?: string;
    workspace?: string;
    page?: number;
    page_size?: number;
  }): Promise<Dataset[]> {
    return this.http.post<Dataset[]>("/datasets/search/roboflow", opts);
  }

  async syncRoboflow(opts: { api_key: string; workspace: string }): Promise<{ status: string }> {
    return this.http.post<{ status: string }>("/datasets/sync/roboflow", opts);
  }
}
